### PR TITLE
[SCB-422] prometheus switch to new mechanism

### PR DESCRIPTION
**Pull request by wujimin**
image

### CODE REVIEW

File: metrics/metrics-integration/metrics-prometheus/src/main/java/org/apache/servicecomb/metrics/prometheus/PrometheusPublisher.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    this.globalRegistry = globalRegistry;
+
+    //prometheus default port allocation is here : https://github.com/prometheus/prometheus/wiki/Default-port-allocations
+    String address =
+        DynamicPropertyFactory.getInstance().getStringProperty(METRICS_PROMETHEUS_ADDRESS, "0.0.0.0:9696").get();
+
+    try {
+      InetSocketAddress socketAddress = getSocketAddress(address);
+      register();
+      this.httpServer = new HTTPServer(socketAddress, CollectorRegistry.defaultRegistry, true);
```
Comment (liubao68) [inline metrics/metrics-integration/metrics-prometheus/src/main/java/org/apache/servicecomb/metrics/prometheus/PrometheusPublisher.java]: For future customization, I think we 'd provide way to export to servicecomb rest framework. And contribute the code to https://github.com/prometheus/client_java

Comment (wujimin) [reply metrics/metrics-integration/metrics-prometheus/src/main/java/org/apache/servicecomb/metrics/prometheus/PrometheusPublisher.java]: By MetricsRestPublisher in previous PR, we already support metrics endpoint: /metrics

and created a issue for contribute: https://issues.apache.org/jira/browse/SCB-466

File: metrics/metrics-integration/metrics-prometheus/pom.xml [hunk 2]
```diff
@@ -43,7 +42,10 @@
       <groupId>org.apache.servicecomb</groupId>
       <artifactId>metrics-core</artifactId>
     </dependency>
-
+    <dependency>
```
Comment (yangbor) [inline metrics/metrics-integration/metrics-prometheus/pom.xml]: Shouldn't the scope set to 'test' here?

Comment (wujimin) [reply metrics/metrics-integration/metrics-prometheus/pom.xml]: why test?
all servicecomb measurement come from metrics-core.

**Comment by yangbor** [comment]
Some files don't have proper final newline.

File: metrics/metrics-integration/metrics-prometheus/pom.xml (hunk 2 above)
Comment (yangbor) [reply metrics/metrics-integration/metrics-prometheus/pom.xml]: I mean 'foundation-test-scaffolding'

File: metrics/metrics-integration/metrics-prometheus/src/test/java/org/apache/servicecomb/metrics/prometheus/TestPrometheusPublisher.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import com.netflix.spectator.api.CompositeRegistry;
+import com.netflix.spectator.api.Counter;
+import com.netflix.spectator.api.DefaultRegistry;
+import com.netflix.spectator.api.ManualClock;
+import com.netflix.spectator.api.Registry;
+import com.netflix.spectator.api.SpectatorUtils;
+import com.sun.net.httpserver.HttpServer;
```
Comment (WillemJiang) [inline metrics/metrics-integration/metrics-prometheus/src/test/java/org/apache/servicecomb/metrics/prometheus/TestPrometheusPublisher.java]: Do we still need to use com.sun package ?

File: metrics/metrics-integration/metrics-prometheus/src/test/java/org/apache/servicecomb/metrics/prometheus/TestPrometheusPublisher.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    ArchaiusUtils.setProperty(PrometheusPublisher.METRICS_PROMETHEUS_ADDRESS, "localhost:0");
+    publisher.init(globalRegistry, null, null);
+
+    Registry registry = new DefaultRegistry(new ManualClock());
+    globalRegistry.add(registry);
+
+    Counter counter = registry.counter("count.name", "tag1", "tag1v", "tag2", "tag2v");
+    counter.increment();
+
+    HTTPServer httpServer = (HTTPServer) FieldUtils.readField(publisher, "httpServer", true);
```
Comment (WillemJiang) [inline metrics/metrics-integration/metrics-prometheus/src/test/java/org/apache/servicecomb/metrics/prometheus/TestPrometheusPublisher.java]: Can we find another way to find the server published address?

Comment (WillemJiang) [reply metrics/metrics-integration/metrics-prometheus/src/test/java/org/apache/servicecomb/metrics/prometheus/TestPrometheusPublisher.java]: Fine.

**Pull request merged**